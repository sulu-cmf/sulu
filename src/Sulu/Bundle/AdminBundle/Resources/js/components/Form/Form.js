// @flow
import React from 'react';
import type {Node} from 'react';
import Grid from '../Grid';
import Field from './Field';
import Section from './Section';
import gridStyles from './grid.scss';

type Props = {|
    children: Node,
|};

export default class Form extends React.Component<Props> {
    static Field = Field;
    static Section = Section;

    render() {
        const {children} = this.props;

        return (
            <Grid className={gridStyles.grid}>
                {children}
            </Grid>
        );
    }
}
