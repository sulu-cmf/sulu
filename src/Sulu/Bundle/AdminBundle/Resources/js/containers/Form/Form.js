// @flow
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import React from 'react';
import Loader from '../../components/Loader';
import Renderer from './Renderer';
import FormStore from './stores/FormStore';

type Props = {
    store: FormStore,
    onSubmit: (action: ?string) => void,
};

@observer
export default class Form extends React.Component<Props> {
    @observable showAllErrors = false;

    /** @public */
    @action submit = (action: ?string) => {
        this.showAllErrors = true;
        this.props.onSubmit(action);
    };

    handleChange = (name: string, value: mixed) => {
        this.props.store.change(name, value);
    };

    handleFieldFinish = () => {
        this.props.store.validate();
    };

    render() {
        const {store} = this.props;

        return store.loading
            ? <Loader />
            : (
                <form>
                    <Renderer
                        data={store.data}
                        errors={store.errors}
                        locale={store.locale}
                        onChange={this.handleChange}
                        onFieldFinish={this.handleFieldFinish}
                        schema={store.schema}
                        showAllErrors={this.showAllErrors}
                    />
                </form>
            );
    }
}
