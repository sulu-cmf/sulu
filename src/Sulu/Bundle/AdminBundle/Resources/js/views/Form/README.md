The form is registered with the key `sulu_admin.form`. It shows a Form with a toolbar. The location to which the back
button redirects can be configured using the route's options. In addition to that it also takes the locales, which
should be shown in the toolbar and an instance of the `ResourceStore`, which allows to load and save data.

| Option                      | Description                                                                           |
|-----------------------------|---------------------------------------------------------------------------------------|
| backRoute                   | The route to which the user will be navigate when the back button is clicked.         |
| editRoute                   | The optional route to which the user will be redirected after the form is saved.      |
| idQueryParameter            | If used the data will be loaded initially from the API by appending this parameter    |
|                             | with the ID.                                                                          |
|                             | It will also be used to be appended when saving a form for the first time.            |
| locales                     | Defines which locales are available in the locale chooser of this form.               |
| resourceStore               | The store which allows to save and load data from the given type of resources         |
| resourceKey                 | Can be set to use another resourceStore than the one being already passed, e.g. for   |
|                             | forms depending on both stores.                                                       |
| routerAttributesToFormStore | Defines which attributes from the [`Router`](#router) are passed to the FormStore.    |
|                             | They will be appended with the same name as in the router to the requests sent from   |
|                             | the `FormStore`, e.g. for saving the form.                                            |
| toolbarAction               | Defines which items registered in the `ToolbarActionRegistry` should be shown in this |
|                             | specific Form.
