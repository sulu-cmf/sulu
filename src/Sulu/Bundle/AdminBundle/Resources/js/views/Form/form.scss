@import '../../containers/Application/variables.scss';

.form {
    padding: $viewPadding;
}
