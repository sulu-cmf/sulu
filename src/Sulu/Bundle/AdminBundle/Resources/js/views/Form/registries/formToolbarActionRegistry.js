// @flow
import AbstractFormToolbarAction from '../toolbarActions/AbstractFormToolbarAction';

class FormToolbarActionRegistry {
    toolbarActions: {[name: string]: Class<AbstractFormToolbarAction>} = {};

    constructor() {
        this.clear();
    }

    clear() {
        this.toolbarActions = {};
    }

    add(name: string, item: Class<AbstractFormToolbarAction>) {
        if (name in this.toolbarActions) {
            throw new Error('The key "' + name + '" has already been used for another ToolbarAction!');
        }

        this.toolbarActions[name] = item;
    }

    get(name: string) {
        if (!(name in this.toolbarActions)) {
            throw new Error('There is no toolbar item with key "' + name + '" registered!');
        }

        return this.toolbarActions[name];
    }
}

export default new FormToolbarActionRegistry();
