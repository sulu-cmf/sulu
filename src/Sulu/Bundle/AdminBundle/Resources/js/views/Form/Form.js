// @flow
import React from 'react';
import {default as FormContainer, FormStore} from '../../containers/Form';
import {withToolbar} from '../../containers/Toolbar';
import type {ViewProps} from '../../containers/ViewRenderer';
import {translate} from '../../utils/Translator';
import ResourceStore from '../../stores/ResourceStore';
import formStyles from './form.scss';

type Props = ViewProps & {
    resourceStore: ResourceStore,
};

class Form extends React.PureComponent<Props> {
    formStore: FormStore;
    form: ?FormContainer;

    componentWillMount() {
        const {resourceStore, router} = this.props;
        this.formStore = new FormStore(resourceStore);

        if (resourceStore.locale) {
            router.bind('locale', resourceStore.locale);
        }
    }

    componentWillUnmount() {
        this.formStore.destroy();
        const {resourceStore, router} = this.props;

        if (resourceStore.locale) {
            router.unbind('locale', resourceStore.locale);
        }
    }

    handleSubmit = () => {
        const {resourceStore, router} = this.props;

        const {
            route: {
                options: {
                    editRoute,
                },
            },
        } = router;

        if (editRoute) {
            resourceStore.destroy();
        }

        this.formStore.save()
            .then(() => {
                if (editRoute) {
                    router.navigate(editRoute, {id: resourceStore.id, locale: resourceStore.locale});
                }
            })
            .catch(() => {
                // TODO show an error label
            });
    };

    setFormRef = (form) => {
        this.form = form;
    };

    render() {
        return (
            <div className={formStyles.form}>
                <FormContainer
                    ref={this.setFormRef}
                    store={this.formStore}
                    onSubmit={this.handleSubmit}
                />
            </div>
        );
    }
}

export default withToolbar(Form, function() {
    const {router} = this.props;
    const {backRoute, locales} = router.route.options;
    const formTypes = this.formStore.types;

    const backButton = backRoute
        ? {
            onClick: () => {
                const {resourceStore} = this.props;

                const options = {};
                if (resourceStore.locale) {
                    options.locale = resourceStore.locale.get();
                }
                router.restore(backRoute, options);
            },
        }
        : undefined;
    const locale = locales
        ? {
            value: this.props.resourceStore.locale.get(),
            onChange: (locale) => {
                this.props.resourceStore.setLocale(locale);
            },
            options: locales.map((locale) => ({
                value: locale,
                label: locale,
            })),
        }
        : undefined;

    const items = [
        {
            type: 'button',
            value: translate('sulu_admin.save'),
            icon: 'su-save',
            disabled: !this.props.resourceStore.dirty,
            loading: this.props.resourceStore.saving,
            onClick: () => {
                this.form.submit();
            },
        },
    ];

    if (this.formStore.typesLoading || Object.keys(formTypes).length > 0) {
        items.push({
            type: 'select',
            icon: 'fa-paint-brush',
            onChange: (value) => {
                this.formStore.changeType(value);
            },
            loading: this.formStore.typesLoading,
            value: this.formStore.type,
            options: Object.keys(formTypes).map((key) => ({
                value: formTypes[key].key,
                label: formTypes[key].title,
            })),
        });
    }

    return {
        backButton,
        locale,
        items,
    };
});
