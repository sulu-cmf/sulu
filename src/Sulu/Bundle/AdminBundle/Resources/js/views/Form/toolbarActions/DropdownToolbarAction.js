// @flow
import React, {Fragment} from 'react';
import {ResourceFormStore} from '../../../containers/Form';
import type {DropdownOption} from '../../../components/Toolbar/types';
import Router from '../../../services/Router';
import formToolbarActionRegistry from '../registries/formToolbarActionRegistry';
import Form from '../Form';
import AbstractFormToolbarAction from './AbstractFormToolbarAction';

export default class DropdownToolbarAction extends AbstractFormToolbarAction {
    toolbarActions: Array<AbstractFormToolbarAction> = [];

    constructor(
        resourceFormStore: ResourceFormStore,
        form: Form,
        router: Router,
        locales: ?Array<string>,
        options: {[key: string]: mixed}
    ) {
        super(
            resourceFormStore,
            form,
            router,
            locales,
            options
        );

        const {toolbarActions} = this.options;

        if (!Array.isArray(toolbarActions)) {
            throw new Error('The passed "toolbarActions" option must be of type object or array');
        }

        this.toolbarActions = toolbarActions
            .map((action) => {
                if (action === null || typeof action !== 'object') {
                    throw new Error('The passed entries in the "actions" option must be objects');
                }

                const {type, options} = action;

                if (typeof type !== 'string') {
                    throw new Error('The "type" of each entry in the "actions" options must be a string');
                }

                if (options === null || typeof options !== 'object') {
                    throw new Error('The "options" of each entry in the "actions" options must be a string');
                }

                return new (formToolbarActionRegistry.get(type))(
                    this.resourceFormStore,
                    this.form,
                    router,
                    this.locales,
                    ((options: any): {[key: string]: mixed})
                );
            });
    }

    getNode() {
        return (
            // TODO Don't hardcode key to allow multiple usage of this action
            <Fragment key="sulu_admin.dropdown">
                {this.toolbarActions.map((toolbarAction) => toolbarAction.getNode())}
            </Fragment>
        );
    }

    getToolbarItemConfig() {
        const {icon, label} = this.options;

        if (typeof label !== 'string') {
            throw new Error('The "label" option must be a string!');
        }

        if (typeof icon !== 'string') {
            throw new Error('The "label" option must be a string!');
        }

        const options: Array<DropdownOption> = this.toolbarActions
            .reduce((toolbarActions, toolbarAction) => {
                const toolbarItemConfig = toolbarAction.getToolbarItemConfig();

                if (!toolbarItemConfig) {
                    return toolbarActions;
                }

                if (toolbarItemConfig.options) {
                    throw new Error('This ToolbarAction only supports child ToolbarActions not being a dropdown');
                }

                const {disabled, label, onClick} = toolbarItemConfig;

                if (!label) {
                    throw new Error('Child ToolbarActions must return a "label"');
                }

                if (!onClick) {
                    throw new Error('Child ToolbarActions must return a "onClick" handler');
                }

                toolbarActions.push({disabled, label, onClick});

                return toolbarActions;
            }, []);

        if (options.length === 0) {
            return undefined;
        }

        return {
            type: 'dropdown',
            label,
            icon,
            options,
        };
    }
}
