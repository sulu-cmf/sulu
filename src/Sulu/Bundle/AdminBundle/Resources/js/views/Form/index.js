// @flow
import Form from './Form';
import formToolbarActionRegistry from './registries/formToolbarActionRegistry';
import AbstractFormToolbarAction from './toolbarActions/AbstractFormToolbarAction';
import DeleteToolbarAction from './toolbarActions/DeleteToolbarAction';
import DropdownToolbarAction from './toolbarActions/DropdownToolbarAction';
import SaveWithPublishingToolbarAction from './toolbarActions/SaveWithPublishingToolbarAction';
import SaveToolbarAction from './toolbarActions/SaveToolbarAction';
import TypeToolbarAction from './toolbarActions/TypeToolbarAction';
import TogglerToolbarAction from './toolbarActions/TogglerToolbarAction';

export default Form;

export {
    formToolbarActionRegistry,
    AbstractFormToolbarAction,
    DeleteToolbarAction,
    DropdownToolbarAction,
    SaveWithPublishingToolbarAction,
    SaveToolbarAction,
    TypeToolbarAction,
    TogglerToolbarAction,
};
