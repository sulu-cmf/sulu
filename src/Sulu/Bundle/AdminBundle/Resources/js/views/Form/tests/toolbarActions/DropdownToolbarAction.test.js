// @flow
import React from 'react';
import {render} from 'enzyme';
import {ResourceFormStore} from '../../../../containers/Form';
import ResourceStore from '../../../../stores/ResourceStore';
import Router from '../../../../services/Router';
import Form from '../../../../views/Form';
import formToolbarActionRegistry from '../../registries/formToolbarActionRegistry';
import DropdownToolbarAction from '../../toolbarActions/DropdownToolbarAction';

jest.mock('../../../../containers/Form/stores/ResourceFormStore', () => (
    class {
        resourceStore;
        constructor(resourceStore) {
            this.resourceStore = resourceStore;
        }

        get dirty() {
            return this.resourceStore.dirty;
        }

        get saving() {
            return this.resourceStore.saving;
        }

        get data() {
            return this.resourceStore.data;
        }
    }
));

jest.mock('../../../../services/Router', () => jest.fn());

jest.mock('../../../../views/Form', () => jest.fn(function() {
    this.submit = jest.fn();
}));

jest.mock('../../registries/formToolbarActionRegistry', () => ({
    get: jest.fn(),
}));

function createDropdownToolbarAction(options = {}) {
    const resourceStore = new ResourceStore('test');
    const resourceFormStore = new ResourceFormStore(resourceStore, 'test');
    const router = new Router({});
    const form = new Form({
        locales: [],
        resourceStore,
        route: router.route,
        router,
    });

    return new DropdownToolbarAction(resourceFormStore, form, router, [], options);
}

test('Return item config with an option for every action in array and skip undefined button', () => {
    const deleteClickSpy = jest.fn();
    const copyClickSpy = jest.fn();

    formToolbarActionRegistry.get.mockImplementation((key) => {
        switch (key) {
            case 'sulu_admin.delete':
                return class {
                    getToolbarItemConfig() {
                        return {
                            label: 'Delete',
                            onClick: deleteClickSpy,
                        };
                    }
                };
            case 'sulu_admin.copy':
                return class {
                    getToolbarItemConfig() {
                        return {
                            label: 'Copy',
                            onClick: copyClickSpy,
                        };
                    }
                };
            case 'sulu_admin.nothing':
                return class {
                    getToolbarItemConfig() {

                    }
                };
        }
    });

    const dropdownToolbarAction = createDropdownToolbarAction({
        icon: 'su-edit',
        label: 'edit',
        toolbarActions: [
            {type: 'sulu_admin.delete', options: {}},
            {type: 'sulu_admin.copy', options: {}},
            {type: 'sulu_admin.nothing', options: {}},
        ],
    });

    expect(dropdownToolbarAction.getToolbarItemConfig()).toEqual({
        icon: 'su-edit',
        label: 'edit',
        options: [
            {
                label: 'Delete',
                onClick: deleteClickSpy,
            },
            {
                label: 'Copy',
                onClick: copyClickSpy,
            },
        ],
        type: 'dropdown',
    });
});

test('Return item config with options passed to child ToolbarActions', () => {
    formToolbarActionRegistry.get.mockImplementation((key) => {
        switch (key) {
            case 'sulu_admin.delete':
                return class {
                    options;

                    constructor(resourceFormStore, form, router, locales, options) {
                        this.options = options;
                    }

                    getToolbarItemConfig() {
                        return {
                            label: this.options.label,
                            onClick: jest.fn(),
                        };
                    }
                };
            case 'sulu_admin.copy':
                return class {
                    options;

                    constructor(resourceFormStore, form, router, locales, options) {
                        this.options = options;
                    }

                    getToolbarItemConfig() {
                        return {
                            label: this.options.title,
                            onClick: jest.fn(),
                        };
                    }
                };
        }
    });

    const dropdownToolbarAction = createDropdownToolbarAction({
        icon: 'su-edit',
        label: 'edit',
        toolbarActions: [
            {type: 'sulu_admin.delete', options: {label: 'Delete'}},
            {type: 'sulu_admin.copy', options: {title: 'Copy'}},
        ],
    });

    expect(dropdownToolbarAction.getToolbarItemConfig()).toEqual({
        icon: 'su-edit',
        label: 'edit',
        options: [
            expect.objectContaining({
                label: 'Delete',
            }),
            expect.objectContaining({
                label: 'Copy',
            }),
        ],
        type: 'dropdown',
    });
});

test('Return no item config if all child ToolbarActions return nothing', () => {
    formToolbarActionRegistry.get.mockImplementation((key) => {
        switch (key) {
            case 'sulu_admin.delete':
                return class {
                    getToolbarItemConfig() {}
                };
            case 'sulu_admin.copy':
                return class {
                    getToolbarItemConfig() {}
                };
        }
    });

    const dropdownToolbarAction = createDropdownToolbarAction({
        icon: 'su-edit',
        label: 'edit',
        toolbarActions: [
            {type: 'sulu_admin.delete', options: {label: 'Delete'}},
            {type: 'sulu_admin.copy', options: {title: 'Copy'}},
        ],
    });

    expect(dropdownToolbarAction.getToolbarItemConfig()).toEqual(undefined);
});

test('Throw error if child ToolbarAction is a dropdown', () => {
    formToolbarActionRegistry.get.mockImplementation((key) => {
        switch (key) {
            case 'sulu_admin.delete':
                return class {
                    getToolbarItemConfig() {
                        return {
                            options: [],
                            type: 'dropdown',
                        };
                    }
                };
        }
    });

    const dropdownToolbarAction = createDropdownToolbarAction({
        icon: 'su-edit',
        label: 'edit',
        toolbarActions: [
            {type: 'sulu_admin.delete', options: {}},
        ],
    });
    expect(() => dropdownToolbarAction.getToolbarItemConfig()).toThrow(/not being a dropdown/);
});

test('Throw error if child ToolbarAction has no onClick handler', () => {
    formToolbarActionRegistry.get.mockImplementation((key) => {
        switch (key) {
            case 'sulu_admin.delete':
                return class {
                    getToolbarItemConfig() {
                        return {
                            label: 'Test',
                        };
                    }
                };
        }
    });

    const dropdownToolbarAction = createDropdownToolbarAction({
        icon: 'su-edit',
        label: 'edit',
        toolbarActions: [
            {type: 'sulu_admin.delete', options: {}},
        ],
    });
    expect(() => dropdownToolbarAction.getToolbarItemConfig()).toThrow(/onClick/);
});

test('Throw error if child Toolbaraction has no label', () => {
    formToolbarActionRegistry.get.mockImplementation((key) => {
        switch (key) {
            case 'sulu_admin.delete':
                return class {
                    getToolbarItemConfig() {
                        return {
                        };
                    }
                };
        }
    });

    const dropdownToolbarAction = createDropdownToolbarAction({
        icon: 'su-edit',
        label: 'edit',
        toolbarActions: [
            {type: 'sulu_admin.delete', options: {}},
        ],
    });
    expect(() => dropdownToolbarAction.getToolbarItemConfig()).toThrow(/label/);
});

test('Return JSX for all child ToolbarActions', () => {
    formToolbarActionRegistry.get.mockImplementation((key) => {
        switch (key) {
            case 'sulu_admin.delete':
                return class {
                    getNode() {
                        return <div className="delete" key="delete" />;
                    }
                };
            case 'sulu_admin.copy':
                return class {
                    getNode() {
                        return <div className="copy" key="copy" />;
                    }
                };
            case 'sulu_admin.nothing':
                return class {
                    getNode() {}
                };
        }
    });

    const dropdownToolbarAction = createDropdownToolbarAction({
        icon: 'su-edit',
        label: 'edit',
        toolbarActions: [
            {type: 'sulu_admin.delete', options: {}},
            {type: 'sulu_admin.copy', options: {}},
            {type: 'sulu_admin.nothing', options: {}},
        ],
    });

    expect(render(dropdownToolbarAction.getNode())).toMatchSnapshot();
});

test('Throw error if toolbarActions are neither an object nor an array', () => {
    expect(() => createDropdownToolbarAction({
        icon: 'su-edit',
        label: 'edit',
        toolbarActions: false,
    })).toThrow(/toolbarActions/);
});
