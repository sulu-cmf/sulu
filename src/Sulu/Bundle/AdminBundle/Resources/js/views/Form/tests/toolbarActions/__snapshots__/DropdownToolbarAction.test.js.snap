// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Return JSX for all child ToolbarActions 1`] = `
Array [
  <div
    class="delete"
  />,
  <div
    class="copy"
  />,
]
`;
